enhancements/cluster-capabilities-api.md

```diff
@@ -0,0 +1,337 @@
+
+While the API is immutable, updating to a new version (upgrade or
+downgrade) is safe because nothing will change the values in the
+`status` block of the resource.
+
+#### Mutable API (future work)
+
+Upgrading the API to add new fields is safe because old consumers will
+ignore them until they are also upgraded.
+
+Adding new values to enum fields is not safe, because there is a
+window of time between when the `capabilities` API is updated and when
+the operators using the API are updated. To avoid issues during that
+time, the controller that manages the capabilities API will wait for
+all `ClusterOperator` resources to have the same version and not
+report their status as progressing, before it copies settings from the
+`spec` fields to the `status` used by consumers. This will ensure that
+the other operators will be upgraded and start consuming the old
+settings. Then, when all of the other operators are updated, the
+settings in the `status` fields will be updated and the other
+operators will re-configure their operands.
+
+For a downgrade, the administrator may need to manually update the
+`capabilities` API to change settings that use new values not
+supported in the older version.
+
+### Version Skew Strategy
+
+Clusters deployed using 4.7 or earlier will not have the capabilities
+CRD or resource in place. We have identified two basic approaches to
+address this:
+
+1. We could provide a library for operators to use to retrieve the
+   "current" capabilities settings. That library could provide error
+   handling, so that when the CRD or CR are not found then default
+   values can be returned in a consistent way.
+2. We could introduce the new CRD and a default CR into a 4.7 z-stream
+   release. The CRD and CR would not be used by any 4.7 operators, but
+   by introducing them early we could ensure that they are present
+   before an upgrade to 4.8 where the operators would expect to be
+   able to find them.
+
+Option 1 appears to be less work initially, but might mean that
+long-lived clusters never have a `capabilities` CR created, which
+would mean we have to maintain the library code indefinitely. Option 2
+would mean more coordinated work initially, and would introduce a
+choke point into the supported upgrade graph. A hybrid approach of
+using option 1 for 4.8 and having an operator
+(`cluster-config-operator`?) in 4.8 create the CR if it is not present
+might be best.
+
+*Input needed to make the right choice.*
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+## Alternatives
+
+### Add fields to existing configuration APIs
+
+We could add the fields we know about today to the
+`infrastructure.config.openshift.io` API, instead of creating a new
+API. This would be less work, initially, but muddies the meaning of
```

Comment:
I don't agree that it muddies the meaning of the API.  It certainly fits within the broad strokes of the existing object, and so from a consistency perspective if "the control plane's HA property" relates to the infrastructure needed for the cluster, it belongs on infrastructure (we are parsimonious about objects, and we are skeptical in general of generic objects with a grab bag of generic fields).

Reply:
with externalized control plane use cases, its also fine that we project the Infrastructure CR in the target cluster, so I kind of prefer not having a new object.

/cc @csrwng